src/engine/structures.rs

```diff
@@ -0,0 +1,156 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use std::collections::HashMap;
+use std::slice::IterMut;
+
+use uuid::Uuid;
+
+use super::engine::{HasName, HasUuid};
+
+
+/// Map UUID and name to T items.
+#[derive(Debug)]
+pub struct Table<T: HasName + HasUuid> {
+    items: Vec<T>,
+    name_map: HashMap<String, usize>,
+    uuid_map: HashMap<Uuid, usize>,
+}
```

Comment:
_This_ is the work of a person who has worked out the theory of bidirectional mappings pretty well, as far as I can tell: https://bidict.readthedocs.io/en/latest/. An important invariant of a bidirectional map is a unique set of keys _and_ a unique set of values. The implementation of ```bidir_map``` does not ensure this property, i.e., it allows insertion of a new item to leave duplicate values. The documentation does not specify what happens on duplicates, but it is pretty obvious from reading the code.